src/parse.c

```diff
@@ -790,6 +829,23 @@ handle_addresses(yaml_document_t* doc, yaml_node_t* node, const void* _, GError*
         prefix_len++; /* skip former '/' into first char of prefix */
         prefix_len_num = g_ascii_strtoull(prefix_len, NULL, 10);
 
+        if (value) {
+            if (!is_ip4_address(addr) && !is_ip6_address(addr))
+                return yaml_error(node, error, "malformed address '%s', must be X.X.X.X/NN or X:X:X:X:X:X:X:X/NN", scalar(entry));
+
+            if (!cur_netdef->address_options)
+                cur_netdef->address_options = g_array_new(FALSE, FALSE, sizeof(addr_option*));
+
+            cur_addr_option = g_new0(addr_option, 1);
+            cur_addr_option->address = g_strdup(scalar(key));
+
+            if (!process_mapping(doc, value, address_option_handlers, error))
+                return FALSE;
+
+            g_array_append_val(cur_netdef->address_options, cur_addr_option);
```

Comment:
I've thought a bit about this, and wanted to get more input before "tainting" this PR with more complex code (maybe I'll push a separate commit on top of this one?).

One "easy" way to do this would be to check if the first element on the `cur_netdef->address_options` array is equal to the struct we're currently pushing there (meaning we're past the first YAML parsing round). This would probably need an additional comparison helper, and we would also need to take into account addresses that are set without specifying any of the options (these would skip this code block, and would need to be handled further down in their IPv4/IPv6 sections).

How is this being handled currently for addresses without any options? How viable would it be to try and detect when the parser needs multiple passes, so that we can just verify if the addresses were set correctly and completely skip the helpers here?

Reply:
Pushing a separate commit on top is a good idea. Keep it in small logical chunks/commits. We will squash all commits from this PR into a single "merge commit" anyways, once it is accepted into master.

It might be even easier to compare the length `cur_netdef->address_options->len`. It will be 0 on first pass, while it will be equal to the length of the YAML sequence in following passes. This is what I did in the other commit (https://github.com/CanonicalLtd/netplan/commit/cc013bdd547f4fcf9743ee8c3ea71f81e7a8d14f):

```
    /* Avoid adding the same peers in a 2nd parsing pass by comparing
       the array size to the YAML sequence size. Skip if they are equal. */
    guint item_count = node->data.sequence.items.top - node->data.sequence.items.start;
    if (cur_netdef->wireguard_peers->len == item_count) {
        g_debug("%s: all wireguard peers have already been added", cur_netdef->id);
        return TRUE;
    }
```

For IP4/IP6 addresses this is currently handled by comparing all IP address strings of the array, which might be a bit overkill... as each parsing pass will contain the very same IP address sequence. And in fact this fix did not yet make it to the `master` branch, but can be seen here: https://github.com/CanonicalLtd/netplan/commit/2f2971b7935c9ff6ae14cd68907cfb87b68472c4 (It was introduced as part of a OVS pull request which are currently still tracked in the `openvswitch-support` branch.

Reply:
I'm not sure the length comparison will work out in this case. If we take the
example below, for instance:
```yaml
network:
  version: 2
  ethernets:
    engreen:
      addresses:
        - 192.168.14.2/24:
            label: out
            lifetime: 0
        - 192.168.14.3/24
        - 2001:FFfe::1/64:
            label: ipv6
```

For this example, `node->data.sequence.items.top - node->data.sequence.items.start` will result in `3` (two IPv4 keys and one IPv6 key), but we'll never reach that amount in `cur_netdef->address_options` as only two of the addresses set options. And I might be missing something here, but I don't think we can know how many addresses set options in the middle of parsing. This seems to be an issue for the IPv4/IPv6 addresses as well, right? (i.e. we can't know how many of the keys will be IPv4 and how many will be IPv6 until we're done parsing all of them).

I've added a draft of how we could do this for address options in a separate commit, but it's not a very elegant solution imho (it also breaks `check-coverage` atm :). If we have no way of knowing that we've done a first pass already, I don't think we'll be able to implement something much different from that, unfortunately. What do you think? Any other ideas?

Reply:
You are absolutely correct! Length comparison won't work out here, due to having MAPPING and SCALAR nodes handled differently.

You draft is looking pretty good already.
Well, I guess we could introduce a new global variable, which increments with each new parsing pass, to know number of the current pass. But this would be very abstract and might confuse people... I think I'd like to keep the multi-pass handling local to this function and array, so it can be easily understood.

Allocating a new `cur_addr_option`, processing all it's mappings, comparing each field inside via `addr_options_equals` and free'ing it up afterwards seems like lots of overhead, though. I think we should move the check prior to allocating new memory for `cur_addr_option` and just compare the `opts->address` VS `scalar(key)`, which we already have at that point. This way we know if there are any options set for this address and can skip it's handling completely.

I prepared a test-case, adding a bridge to trigger multi-pass parsing. Also, I drafted my thoughts from above in this commit: https://github.com/slyon/netplan/commit/1734de6ec339be5219fab55a439c17c7cee47f51

Feel free to cherry-pick and/or adopt.